tcpproxy_test.go

```diff
@@ -0,0 +1,78 @@
+// Copyright 2015 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+package testing_test
+
+import (
+	"fmt"
+	"io"
+	"net"
+	"sync"
+
+	"github.com/juju/testing"
+	gc "gopkg.in/check.v1"
+)
+
+var _ = gc.Suite(&tcpProxySuite{})
+
+type tcpProxySuite struct{}
+
+func (*tcpProxySuite) TestTCPProxy(c *gc.C) {
+	var wg sync.WaitGroup
+
+	listener, err := net.Listen("tcp", "127.0.0.1:0")
+	c.Assert(err, gc.IsNil)
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		conn, err := listener.Accept()
+		if err != nil {
+			c.Errorf("cannot accept: %v", err)
+			return
+		}
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			defer conn.Close()
+			// Echo anything that was written.
```

Comment:
I would have thought an echo server might be generally useful perhaps break that out as a thing too?